lib/spack/spack/test/installer.py

```diff
@@ -1177,3 +1178,97 @@ def test_install_skip_patch(install_mockery, mock_fetch):
 
     spec, install_args = const_arg[0]
     assert inst.package_id(spec.package) in installer.installed
+
+
+def test_overwrite_install_backup_success(temporary_store, config, mock_packages,
+                                          tmpdir):
+    """
+    When doing an overwrite install that fails, Spack should restore the backup
+    of the original prefix, and leave the original spec marked installed.
+    """
+    # Where to store the backups
+    backup = str(tmpdir.mkdir("backup"))
+
+    # Get a build task (TODO: can this be done without calling internal methods?)
```

Comment:
One way to simplify this might be to adjust the API of `OverwriteInstall`:

```
class OverwriteInstall(object):
  def __init__(self, task, fail_action):
    self.task = task
    self.fail_action = fail_action

  def install(self):
    with replace_directory_transation:
      task()
    except fs.CouldNotRestoreDirectoryBackup as e:
      fail_action
      ...
      raise e.inner_exception
```

That would also remove the need for a mock DB etc.

I think this is good the way it is though, so I wouldn't require this change (I'm primarily brought it up since you included a question in the comment).

Reply:
Yeah, something along those lines could work too, but generally I kinda like dependency injection like I've done it here, even though Python doesn't really have interfaces (?) and making classes and anonymous classes (aka the lambda's of classes :p).

Reply:
My main issue was that getting a build task is either very verbose, or it calls into the private API of the installer (judging from the underscore prefix that is)

Reply:
Done! Thanks!